src/network/packets/clientbound.rs

```diff
@@ -8,230 +8,236 @@ pub trait ClientBoundPacket {
-    pub animation: u8,
-}
-
-impl ClientBoundPacket for CEntityAnimation {
-    fn encode(self) -> PacketEncoder {
-        let mut buf = Vec::new();
-        buf.write_varint(self.entity_id);
-        buf.write_unsigned_byte(self.animation);
-        PacketEncoder::new(buf, 0x05)
-    }
-}
-
-pub struct CBlockEntityData {
-    pub x: i32,
-    pub y: i32,
-    pub z: i32,
-    pub action: u8,
-    pub nbt: nbt::Blob,
-}
-
-impl ClientBoundPacket for CBlockEntityData {
-    fn encode(self) -> PacketEncoder {
-        let mut buf = Vec::new();
-        buf.write_position(self.x, self.y, self.z);
-        buf.write_unsigned_byte(self.action);
-        buf.write_nbt_blob(self.nbt);
-        PacketEncoder::new(buf, 0x09)
-    }
-}
-
-pub struct CBlockChange {
-    pub x: i32,
-    pub y: i32,
-    pub z: i32,
-    pub block_id: i32,
-}
+macro_rules! clientbound_packets {
+    (
+        $(
+            $(#[$($struct_meta:meta),+])?
+            $name:ident -> $packet_id:literal {
+                $(pub $pfield:ident: $pfield_ty:ident,)*
+                $(fn $field_write:ident($($field:ident: $field_ty:path),+) => ($($field_expr:expr),+),)*
+            },
+        )*
+    ) => {
+        $(
+            $(#[$($struct_meta),+])?
+            pub struct $name {
+                $($(pub $field: $field_ty, )*)*
+                $(pub $pfield: $pfield_ty, )*
+            }
 
-impl ClientBoundPacket for CBlockChange {
-    fn encode(self) -> PacketEncoder {
-        let mut buf = Vec::new();
-        buf.write_position(self.x, self.y, self.z);
-        buf.write_varint(self.block_id);
-        PacketEncoder::new(buf, 0x0B)
-    }
-}
+            impl ClientBoundPacket for $name {
+                fn encode(self) -> PacketEncoder {
+                    let mut buf = Vec::new();
+                    $( let $pfield = self.$pfield; )*
+                    $($( let $field = self.$field; )*)*
 
-pub struct CChatMessage {
-    pub message: String,
-    pub position: i8,
-    pub sender: u128,
-}
+                    $( buf.$field_write($($field_expr),*);)*
 
-impl ClientBoundPacket for CChatMessage {
-    fn encode(self) -> PacketEncoder {
-        let mut buf = Vec::new();
-        buf.write_string(32767, &self.message);
-        buf.write_byte(self.position);
-        buf.write_uuid(self.sender);
-        PacketEncoder::new(buf, 0x0E)
-    }
-}
+                    PacketEncoder::new(buf, $packet_id)
+                }
+            }
+        )*
+    };
+}
+
+clientbound_packets!(
+    // Server List Ping Packets
+    CResponse -> 0x00 {
+        fn write_string(json_response: String) => (32767, &json_response),
+    },
+
+    // Login Packets
+    CDisconnectLogin -> 0x00 {
+        fn write_string(reason: String) => (32767, &reason),
+    },
+    CPong -> 0x01 {
+        fn write_long(payload: i64) => (payload),
+    },
+    CLoginSuccess -> 0x02 {
+        fn write_uuid(uuid: u128) => (uuid),
+        fn write_string(username: String) => (16, &username),
```

Comment:
Maybe instead of commas here use semicolons? 

```suggestion
        fn write_uuid(uuid: u128) => (uuid);
        fn write_string(username: String) => (16, &username);
```